### ISSUE TITLE
Maybe it's better than #1541. What do you think?

### ISSUE THREAD

**Issue by sephid86** [wwmm/easyeffects#1548]
pipe_manager.cpp
~  243 //    if (nd->nd_info->name == "OBS") {
   244       if (g_strcmp0(spa_dict_lookup(info->props, PW_KEY_STREAM_CAPTURE_SINK), "true") == 0) {
~  245         nd->nd_info->capture_sink=spa_dict_lookup(info->props, PW_KEY_STREAM_CAPTURE_SINK);
~  246 //        remove_node = true;
+  247 //      }
   248     }

pipe_manager.hpp
   63   std::string application_id;
   64 
+  65   std::string capture_sink;
+  66 

app_info.cpp
  216 void update(AppInfo* self, const NodeInfo node_info) {
  217   if (node_info.state == PW_NODE_STATE_CREATING) {
  218     // PW_NODE_STATE_CREATING is useless and does not give any meaningful info, therefore skip it
  219     return;
  220   }
  221   
  222   self->data->info = node_info;
+ 223   if(node_info.capture_sink == "true") {
+ 224   self->data->application->pm->disconnect_stream(node_info.id);
+ 225   }

Maybe it's better than #1541. What do you think?

Through testing, I confirmed that the intended function works.

I don't think this is complete.

In the #1541 code I posted before, 'desktop audio' is not displayed on the easyeffects screen, but when 'process all input' is on in preference

I could see in pavucontrol that 'desktop audio' is connected to the easyeffects source.

The code above shows 'desktop audio' on the easyeffects screen but doesn't allow 'desktop audio' to connect to the easyeffects source.

You know more about easyeffects than I do, so I think you can implement it in a better way than this.

I don't know C++ language at all. I've never dealt with it.

I know very little C language. I can do a little. beginner level.

So it was very difficult to modify the source code to work the intended function in easyeffects.

It was a tough two days.

I hope this conveys a little bit of what I mean.

Thank you for your kind attention. sir.

I'm going to go make my own program that I'm making now.

I don't know if my lack of skills can be of any help, but if you contact me at any time, I will reply.

테스트를 통해 내가 의도한 기능이 작동하는것을 확인하였습니다.

이것이 완전하다고 생각하지 않습니다.

전에 올렸던 #1541 코드는 'desktop audio'가 easyeffects 화면상에 표시되지 않지만 preference 에서 'process all input' 이 on 일때

'desktop audio' 가 easyeffects source 에 연결되어 있음을 pavucontrol 에서 확인할수 있었습니다.

위의 코드는 easyeffects 화면상에 'desktop audio' 가 표시되지만 'desktop audio' 가 easyeffects source 와 연결되지 않게 해줍니다.

선생님께서는 나보다 easyeffects 에 대해 더 잘 알고 계시니 이것보다 더 좋은 방법으로 구현하실수 있다고 생각합니다.

저는 C++언어를 전혀 모릅니다.그것을 다루어 본적도 없습니다.

C언어는 아주 조금 압니다. 조금 할줄 압니다. 초보자의 수준입니다.

그래서 easyeffects 에서 내가 의도한 기능을 작동하도록 소스코드를 수정 하기가 매우 어려웠습니다.

힘든 이틀이었습니다..

이것으로 제가 의도하는 바가 무엇인지 조금이나마 전달 되기를 바랍니다.

선생님의 친절한 관심에 감사드립니다.

저는 이제 만들고 있던 저의 프로그램을 만들러 가보겠습니다.

혹시 저의 부족한 실력이 도움이 될수 있을지 모르겠지만 언제든 연락을 주시면 답장을 하겠습니다.

**Comment by sephid86** [comment]
a........

igeos-eun easyeffects chang-i hwalseonghwa doeeoissji anh-eumyeon soyong-i eobsneyo...

start login service at login i on ilttaedo...

eolyeobda...

naneun daleun bangbeob-eul chaj-aya gessneyo..
ah........

This is of no use unless the easyeffects window is active...

Even when start login service at login is on...

it's difficult...

I'll have to find another way...

**Comment by wwmm** [comment]
>In the https://github.com/wwmm/easyeffects/issues/1541 code I posted before, 'desktop audio' is not displayed on the easyeffects screen, but when 'process all input' is on in preference

I am doing some tests and I think I see why this is happening. Again the problems is the fact `PW_KEY_STREAM_CAPTURE_SINK` is only set by PipeWire some time after it notifies us about the stream creation. By the time we have this information `StreamInputEffects::on_app_added` has already been called. What makes the stream to be moved to our sources even if it is not shown in our recorders lists... Handling this will be a little annoying...

@Digitalone1 I think that the workaround inside `on_node_info` will have to reset the target metadata in the stream besides removing them from our node map. With luck PipeWire's media session manager will restore the stream to where it should be. I will test this now.

**Comment by wwmm** [comment]
>If we are luck PipeWire's media session manager will restore the stream to where it should be. I will test this now.

We were lucky :smile:. @sephid86 I have updated the master branch. Let me know how it is on your computer. At least on mine now the OBS `desktop stream` is restored to the soundcard monitor.

**Comment by Digitalone1** [comment]
Okay, I see in the commit, the metadata is removed anytime the node is removed from our map.

**Comment by Digitalone1** [comment]
> > Has this to be done everytime the node is removed from the map or only in the specific case of capture stream exclusion?
> 
> For now it is done only when the condition `remove_node` is true. As most programs won't activate this code path it should be fine.

Okay.

**Comment by Digitalone1** [comment]
So the next step is adding a specific user config to enable/disable this task?

**Comment by wwmm** [comment]
>So the next step is adding a specific user config to enable/disable this task?

Yes. It will be better.

**Comment by sephid86** [comment]
> > 운이 좋다면 PipeWire의 미디어 세션 관리자가 스트림을 원래 있던 위치로 복원합니다. 나는 지금 이것을 테스트할 것이다.
> 
> 우리는 운이 좋았다웃다.@sephid86마스터 브랜치를 업데이트했습니다. 컴퓨터에서 어떻게 되는지 알려주세요. 적어도 지금은 OBS `desktop stream`가 사운드 카드 모니터로 복원됩니다.

thank you.

I tested it and it works fine.

I'll format my computer, reinstall linux, and test it again in a clean state, and then I'll tell you the result.

**Comment by sephid86** [comment]
> > So the next step is adding a specific user config to enable/disable this task?
> 
> Yes. It will be better.
![image_3990014981494827719999](https://user-images.githubusercontent.com/77107998/170541975-f63e9c8d-0e5f-4e9e-b8e4-9934a4a62e6d.png)

oh.. captain.. oh... my captain...

Oh... you are the light and salt of the world. sir.


I looked at the modified code before 'format'.

I tested a different method instead of 'format' and it was confirmed that it also works well.

You fixed what I said was a bug in easyeffects.

Now easyeffects doesn't have the bug I was talking about.

What I said were bugs in easyeffects is now resolved.

**I bow my head and say thank you.**

98% satisfied.

Because I am concerned about the situation we did not expect.

2% is that.

I'm trying to figure out how to solve that 2%.

I'm going to try a few things and I'll tell you what the 2% means I'm talking about, whether I've found a way or not.
